cmd/kubeadm/app/cmd/join.go

```diff
@@ -245,27 +296,181 @@ func NewJoin(cfgPath string, args []string, defaultcfg *kubeadmapiv1alpha3.JoinC
 
 // Run executes worker node provisioning and tries to join an existing cluster.
 func (j *Join) Run(out io.Writer) error {
-
 	// Perform the Discovery, which turns a Bootstrap Token and optionally (and preferably) a CA cert hash into a KubeConfig
 	// file that may be used for the TLS Bootstrapping process the kubelet performs using the Certificates API.
-	glog.V(1).Infoln("[join] retrieving KubeConfig objects")
-	cfg, err := discovery.For(j.cfg)
+	glog.V(1).Infoln("[join] discovering cluster-info")
+	tlsBootstrapCfg, err := discovery.For(j.cfg)
+	if err != nil {
+		return err
+	}
+
+	// If the node is joining as a master
+	if j.cfg.Master == true {
+		// Retrives the kubeadm configuration used during kubeadm init
+		glog.V(1).Infoln("[join] retrieving KubeConfig objects")
+		masterConfiguration, err := j.FetchInitClusterConfiguration(tlsBootstrapCfg)
+		if err != nil {
+			return err
+		}
+
+		// injects into the kubeadm configuration used for init the information about the joining node
+		masterConfiguration.NodeRegistration = j.cfg.NodeRegistration
+		masterConfiguration.API.AdvertiseAddress = j.cfg.AdvertiseAddress
+
+		// Checks if the cluster configuration supports
+		// joining the additional master node and if all the necessary certificates are provided
+		if err = j.CheckIfReadyForJoinAsMaster(masterConfiguration); err != nil {
+			// outputs the not ready for join master message
+			ctx := map[string]string{
+				"Error": err.Error(),
+			}
+
+			var msg bytes.Buffer
+			notReadyToJoinMasterTemp.Execute(&msg, ctx)
```

Comment:
I kind of think we should wrap this to return just an error.  Not a blocker for this PR.